### PR TITLE
Cortex M3 arch

### PR DESCRIPTION
**Pull request by cpluss**
### Pull Request Overview

Adds basic support for Cortex-M3 arch. Everything is basically a duplicate of the M4 architecture.

### CODE REVIEW

**Comment by alevy** [comment]
I don't see a reason not to merge this. The stm32 port can switch to using this instead of it's own `cortex-m3` crate.

**Comment by ppannuto** [comment]
First -- I agree, happy to merge this it looks fine and this next question shouldn't stand in the way of progress.

---

Are there _any_ differences between the m3 and m4 crates? It's not stellar to me to just have two copies sitting around.

To that end, perhaps we should be calling this by their architectural names, rather than their marketing names, that is:

  - `cortex-m0` -> `arm-v6m`
  - `cortex-m3` -> `arm-v7m`
  - `cortex-m4` -> `arm-v7me`

These express the contents a little better, as the difference between the current m0/m4 crates is the use of instructions that are in `v7m` that are not in `v6m`, and we're going to get more duplication as time goes on (i.e. the cortex-m0+, which replaced the m0, is also a `v6m` core).

File: arch/cortex-m3/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+     *  */
+	str	r0, [r3, r2, lsl #2]"
+    );
+}
+
+#[no_mangle]
+#[naked]
+#[allow(non_snake_case)]
+pub unsafe extern "C" fn SVC_Handler() {
+    asm!(
```
Comment (niklasad1) [inline arch/cortex-m3/src/lib.rs]: I still think this function should be `lowercase` and remove ```#[allow(non_snake_case)]```

**Comment by niklasad1** [comment]
@ppannuto M3 and M4 are identical implementation-wise but code duplication is okay IMO. Yes, I agree to use v6m, v7, v7em, v8 over cortex-m0, m0+, m1, ... m33. 

Otherwise, I think it's also possible to migrate to one crate such as coretex M and use different features/configs for things that differ such as instructions, alignment and so on. 
It eliminates code duplication but becomes probably harder to maintain.

**Pull request merged**